### PR TITLE
Added new feature for nice error messages to ParamFetcher

### PR DESCRIPTION
**Pull request by feridmovsumov**
fix for https://github.com/FriendsOfSymfony/FOSRestBundle/issues/1013

Example Usage: 
```php
@QueryParam(name="parent_id", requirements={"rule" = "\d+", "error_message" = "parent_id must be an integer"}, strict=true, nullable=true, description="Parent Id")
```

Error Message:
```json
{
    "code": 400,
    "message": "parent_id must be an integer"
}
```

### CODE REVIEW

File: Request/ParamFetcher.php [hunk 1]
```diff
@@ -174,6 +174,7 @@ public function get($name, $strict = null)
         return $this->cleanParamWithRequirements($config, $param, $strict);
     }
 
+
```
Comment (lsmith77) [inline Request/ParamFetcher.php]: please remove this new line

**Comment by lsmith77** [comment]
there are some minor coding style issues but we can get that fixed quickly.
I wonder why did you need to change the method so much? wouldn't it have been enough to just make the sprintf message configurable?

**Comment by feridmovsumov** [comment]
I think this will match requirements of many developers.
```php
requirements={"rule" = "\d+", "message" = "parent_id must be an integer"}
```

**Comment by mvrhov** [comment]
I  think `error_message` would be better to `message`

**Comment by feridmovsumov** [comment]
@mvrhov yes you right. I refactored code.

File: Request/ParamFetcher.php [hunk 2]
```diff
@@ -195,25 +195,22 @@ public function cleanParamWithRequirements(Param $config, $param, $strict)
             );
         }
 
-        $this->checkNotIncompatibleParams($config);
```
Comment (lsmith77) [inline Request/ParamFetcher.php]: the fact that your PR removes this line makes it seem to me like maybe some of the changes you did are just the result of resolving some conflicts during a rebase?

Comment (feridmovsumov) [inline Request/ParamFetcher.php]: I accidentally removed this line. I put it back in my last commit

File: Request/ParamFetcher.php [hunk 3]
```diff
@@ -223,10 +222,15 @@ public function cleanParamWithRequirements(Param $config, $param, $strict)
                     $config->requirements
                 ),
             ));
+        }else{
```
Comment (lsmith77) [inline Request/ParamFetcher.php]: needs to be ``} else {``

File: Request/ParamFetcher.php [hunk 4]
```diff
@@ -237,9 +241,13 @@ public function cleanParamWithRequirements(Param $config, $param, $strict)
 
         if (0 !== count($errors)) {
             if ($strict) {
-                throw new BadRequestHttpException($this->violationFormatter->formatList($config, $errors));
+                if(isset($config->requirements["error_message"])){
+                    $errorMessage = $config->requirements["error_message"];
+                }else{
```
Comment (lsmith77) [inline Request/ParamFetcher.php]: needs to be } else {

File: Request/ParamFetcher.php [hunk 5]
```diff
@@ -237,9 +241,13 @@ public function cleanParamWithRequirements(Param $config, $param, $strict)
 
         if (0 !== count($errors)) {
             if ($strict) {
-                throw new BadRequestHttpException($this->violationFormatter->formatList($config, $errors));
+                if(isset($config->requirements["error_message"])){
```
Comment (lsmith77) [inline Request/ParamFetcher.php]: needs to be ``if (``

File: Request/ParamFetcher.php [hunk 6]
```diff
@@ -203,17 +203,16 @@ public function cleanParamWithRequirements(Param $config, $param, $strict)
 
         $constraint = $config->requirements;
 
-        if (is_scalar($constraint)) {
```
Comment (lsmith77) [inline Request/ParamFetcher.php]: why did you move this ``is_scalar()`` call?

Comment (feridmovsumov) [inline Request/ParamFetcher.php]: Actually I didn't move this code I moved code below

```php
if (is_array($param)) {
            if ($strict) {
                throw new BadRequestHttpException(
                    sprintf("%s parameter is an array", $paramType)
                );
            }
            return $default;
        }
```

**Comment by feridmovsumov** [comment]
I fixed broken tests.

File: .gitignore [hunk 7]
```diff
@@ -2,3 +2,4 @@ phpunit.xml
 vendor
 composer.lock
 composer.phar
+.idea/
```
Comment (xabbuh) [inline .gitignore]: This should be reverted (you can ignore that globally on your system or per project).

File: Request/ParamFetcher.php [hunk 8]
```diff
@@ -204,13 +204,13 @@ public function cleanParamWithRequirements(Param $config, $param, $strict)
         $constraint = $config->requirements;
 
         if (is_scalar($constraint)) {
+
```
Comment (xabbuh) [inline Request/ParamFetcher.php]: The changes to the blank lines should imho be reverted as well.

File: Request/ParamFetcher.php [hunk 9]
```diff
@@ -223,10 +223,15 @@ public function cleanParamWithRequirements(Param $config, $param, $strict)
                     $config->requirements
                 ),
             ));
+        } elseif(is_array($config->requirements) && isset($config->requirements["rule"]) && $config->requirements["error_message"]) {
```
Comment (xabbuh) [inline Request/ParamFetcher.php]: There should be a space after `elseif`.

Comment (xabbuh) [inline Request/ParamFetcher.php]: The `is_array($config->requirements)` check is not necessary.

Comment (feridmovsumov) [inline Request/ParamFetcher.php]: It is required because tests gives an fatal error
```json
PHP Fatal error:  Cannot use object of type Mock_Constraint_d66ec56b as array in /home/web/project/FOSRestBundle/Request/ParamFetcher.php on line 245
```

**Comment by feridmovsumov** [comment]
@xabbuh thanks for code review. I fixed.

**Comment by tystr** [comment]
:+1:

File: Request/ParamFetcher.php [hunk 10]
```diff
@@ -223,10 +222,15 @@ public function cleanParamWithRequirements(Param $config, $param, $strict)
                     $config->requirements
                 ),
             ));
+        } elseif (is_array($config->requirements) && isset($config->requirements["rule"]) && $config->requirements["error_message"]) {
```
Comment (lsmith77) [inline Request/ParamFetcher.php]: you can use ``$constraint`` here instead of ``$config->requirements``

File: Request/ParamFetcher.php [hunk 11]
```diff
@@ -223,10 +222,15 @@ public function cleanParamWithRequirements(Param $config, $param, $strict)
...
+            $constraint = new Regex(array(
+                'pattern' => '#^'.$config->requirements["rule"].'$#xsu',
+                'message' => $config->requirements["error_message"]
+            ));
+        }
 
-            if (false === $config->allowBlank) {
-                $constraint = array(new NotBlank(), $constraint);
-            }
+        if (false === $config->allowBlank) {
```
Comment (lsmith77) [inline Request/ParamFetcher.php]: moving this out of the if introduced a new code path, can you check that this doesn't have any adverse side effects?

File: Request/ParamFetcher.php (hunk 10 above)
Comment (feridmovsumov) [inline Request/ParamFetcher.php]: ok I fixed that.

File: Request/ParamFetcher.php (hunk 11 above)
Comment (feridmovsumov) [inline Request/ParamFetcher.php]: There is no any risk I tested it.

File: Request/ParamFetcher.php [hunk 12]
```diff
@@ -210,7 +210,6 @@ public function cleanParamWithRequirements(Param $config, $param, $strict)
                         sprintf("%s parameter is an array", $paramType)
                     );
                 }
-
```
Comment (xabbuh) [inline Request/ParamFetcher.php]: I would revert this change.

File: Request/ParamFetcher.php [hunk 13]
```diff
@@ -237,9 +241,13 @@ public function cleanParamWithRequirements(Param $config, $param, $strict)
...
             if ($strict) {
-                throw new BadRequestHttpException($this->violationFormatter->formatList($config, $errors));
+                if (is_array($config->requirements) && isset($config->requirements["error_message"])) {
+                    $errorMessage = $config->requirements["error_message"];
+                } else {
+                    $errorMessage = $this->violationFormatter->formatList($config, $errors);
+                }
+                throw new BadRequestHttpException($errorMessage);
             }
-
```
Comment (xabbuh) [inline Request/ParamFetcher.php]: and this as well

**Comment by stof** [comment]
This new features should be covered in the testsuite before being accepted

**Comment by feridmovsumov** [comment]
I will add tests soon.

**Comment by feridmovsumov** [comment]
I am trying to add tests for my changes but I am getting following error when I run phpunit. I didn't get it before.
```json
Configuration read from /home/web/project/FOSRestBundle/phpunit.xml.dist

...........PHP Fatal error:  Call to undefined method Mock_ContainerBuilder_210a2f5e::method() in /home/web/project/FOSRestBundle/Tests/DependencyInjection/Compiler/SerializerConfigurationPassTest.php on line 48
```

I am running phpunit tests with command below in root folder
```
phpunit .
```
@stof @xabbuh @lsmith77 can you help me please

**Comment by feridmovsumov** [comment]
@stof @lsmith77 I added tests for custom error messages.

**Comment by lsmith77** [comment]
looks good now I guess ..

can you make those minor changes that @xabbuh noted?

**Comment by feridmovsumov** [comment]
@lsmith77 I removed unnecessary lines.

**Comment by lsmith77** [comment]
can you address the two comments by @xabbuh about removed empty lines?

**Comment by feridmovsumov** [comment]
@lsmith77 I fixed line problem mentioned by @xabbuh You can see lines below

https://github.com/feridmovsumov/FOSRestBundle/blob/master/Request/ParamFetcher.php#L250

https://github.com/feridmovsumov/FOSRestBundle/blob/master/Request/ParamFetcher.php#L213

**Comment by lsmith77** [comment]
hmm .. I do not see it on my end .. anyway .. will merge and tweak if necessary

**Comment by ericc06** [comment]
Hi,
I tried to use the QueryParam custom error message, but I can make it work only in debug mode. More details [here](https://github.com/FriendsOfSymfony/FOSRestBundle/issues/2086).

How can we send in the JSON response the custom error message in production mode (i.e. when APP_DEBUG=0)?
